NIFI-3572 Add PipelineXml processor to support XProc XML pipelines via XMLCalabash

stevedlawrence opened the pull request:
This Pull Request is based off of a repo on NCSA server:

https://opensource.ncsa.illinois.edu/bitbucket/projects/DFDL/repos/nifi-xproc/browse

This was sent to the apache dev mailing list here:

https://mail-archives.apache.org/mod_mbox/nifi-dev/201703.mbox/%3C5d4c2d48-79de-5543-c704-27059c2644ff%40tresys.com%3E

Feedback was incorporated and changes where  squashed and files moved to the appropriate locations in the nifi repo.

----

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [X] Is your initial contribution a single, squashed commit?

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [X] Have you written or updated unit tests to verify your changes?
- [X] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [X] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [X] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [X] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [X] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

stevedlawrence commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PipelineXml.java:
```diff
@@ -0,0 +1,451 @@
+            .addValidator(Validator.VALID)
+            .addValidator(XML_PIPELINE_VALIDATOR)
+            .build();
+    public static final PropertyDescriptor XML_PIPELINE_POOL_SIZE = new PropertyDescriptor.Builder()
```
This pool size configuration option is really not necessary. Any size over the number of concurrent tasks will lead to pool instances never being used and just wasting memory. And any size under the number of concurrent tasks will effectively limit how many concurrent tasks can run at once. So instead of making this configurable, the pool size should just be initialized to the maximum number of concurrent tasks.

stevedlawrence replied:
A new commit was added to this pull request to make this change.

stevedlawrence commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PipelineXml.java:
```diff
@@ -0,0 +1,451 @@
+        try {
+            pd = pipelinePool.take();
+        } catch (InterruptedException e) {
+            return;
```
I suspect instead of returning, this should call Thread.currentThread().interrupt()

stevedlawrence commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PipelineXml.java:
```diff
@@ -0,0 +1,451 @@
+
+        final FlowFile original = session.get();
+        if (original == null) {
+            return;
```
If there was no FlowFile, then we never give the pipeline back to the pipeline pool, thus shrinking the size of the pool. This should put it back so other threads can use it.

trixpan commented:
@stevedlawrence seems like the PR is not a squashed PR. mind rebasing?

stevedlawrence commented:
@trixpan Sure thing. Now squashed and rebased onto the latest origin/master.

stevedlawrence commented:
Note, the commit message in the first squash (cd6aed3) included messages from the squashed commits. I've force pushed a new commit (3cc426a) with a correct commit message. The code in the two commits are the same.

joewitt commented:
hello.  Would definitely prefer/recommend we put this into its own nar bundle and not along side the standard processors bundle.  'nifi-xmlcalabash-bundle' for example.  Are there any templates/examples of use to leverage for testing?

stevedlawrence commented:
Changes made to latest pull request:
- Rebased onto latest master, and updated to 1.3.0-SNAPSHOT
- Moved the pipelinexml processor out of standard bundle and into nifi-xmlcalabash-bundle
- Add new Base URI option, which is needed in some cases when providing the XProc config as a string rather than a File
- Minor fixes based on -Pcontrib-check

Also, I've attached a template that shows a very simple use of the PipelineXML processor. It's probably not that realistic, but does show some more complicated apsects like auto detecting the output relationships based on the XProc config and dealing with XProc ports with multiple output files for a single port. The template includes a description of what it does in a bit more detail.
[PipelineXML_Example.xml.zip](https://github.com/apache/nifi/files/1032779/PipelineXML_Example.xml.zip)

joewitt commented:
Closing due to inactivity.
